proto/keyserverconfig.proto

```diff
@@ -119,24 +119,114 @@ message KeyserverConfig {
+	// this keyserver accepts email address registrations.
 	repeated string email_proof_allowed_domains = 10;
 
 	bool insecure_skip_email_proof = 11;
+
+	// Keyserver may support multiple registration policies at a time.
+	// A policy is acceptable only for the domains it supports.
+	repeated RegistrationPolicy registration_policy = 12;
+}
+
+// RegistrationPolicy specifies the list of policies the keyserver may support,
+// to confirm the ownership of an email address
+message RegistrationPolicy {
+	oneof policy_type {
+		EmailProofByDKIM email_proof_by_dkim = 1	[(gogoproto.customname) = "EmailProofByDKIM"];
+		EmailProofByClientCert email_proof_by_client_cert = 2;
+		EmailProofByOIDC email_proof_by_oidc = 3	[(gogoproto.customname) = "EmailProofByOIDC"];
+		EmailProofByCustomAuth email_proof_by_custom_auth = 4;
+	}
+}
+
+message EmailProofByDKIM {
+	// AllowedDomains specifies the domains for which this keyserver accepts
+	// email address registrations.
+	repeated string allowed_domains = 1;
+	// ToAddr specifies the additional allowed to address in email proofs. By
+	// default, only proofs sent to the user being registered all accepted.
+	// This option can be used to allow proofs emailed directly to the
+	// keyserver to be accepted (but the keyserver does NOT implement a SMTP
+	// interface right now).
+	string to_addr = 2;
+	// SubjectPrefix is used for DKIM-based email address registration.  The
+	// proof challenge is sent in the subject line, with an optional string
+	// preceding it. For example, if EmailProofSubjectPrefix = "account
+	// verification: ", then the proof email needs to have a subject line
+	// "account verification: abcd" for verify challenge abcd.
+	string subject_prefix = 3;
+}
+
+// EmailProofByClientCert accepts a certificate signed by an authority trusted
+// with handling registration as sufficient confirmation of ownership of an
+// email address. The emailAddress value in the certificate's DistinguishedName
+// field is allowed to be regeistered by the holder of the key specified in the
+// certificate.
+message EmailProofByClientCert {
+	// AllowedDomains specifies the domains for which this keyserver accepts
+	// email address registrations by this policy.
+	repeated string allowed_domains = 1;
+	// CaCert contains the trusted certificate authority certificate in DER format.
+	bytes ca_cert = 2;
+}
+
+// EmailProofByOIDC accepts an ID token fetched from an OpenID Connect provider
+// and validates it as specified in the configuration. The token must have a 
+// valid signature, not expired, signed by the expected provider and must contain
+// the expected Client ID
+message EmailProofByOIDC {
+	repeated OIDCConfig oidc_config = 1 [(gogoproto.customname) = "OIDCConfig"];
+}
+
+// EmailProofByCustomAuth accepts opaque bytes which keyserver can use to implement
+// a custom verification to confirm ownership of an email address.
+message EmailProofByCustomAuth {
```

Comment:
If this is removed from RegistrationPolicy as I suggested, also delete it from here :).